### DISCUSSION TITLE
Count the number of triangular faces

### DISCUSSION THREAD

**Discussion by peiyanlu** [iTwin/itwinjs-core#7431] [Q&A]
Hi, I've been conducting performance statistics recently. In other engines, the number of triangular faces is often used to quantify the size of a model, serving as one of the indicators that affect performance. Therefore, I would like to inquire if there is a method in iTwin to calculate the number of triangular faces a model has after it is published.

**Comment by pmconne** [comment]
Hi @peiyanlu,

iModels generally contain vector geometry, not triangle meshes. To visualize an iModel, we produce from the vector geometry [3D tilesets](https://www.ogc.org/publications/standard/3dtiles/) containing graphical primitives like triangles. Because 3D tilesets employ hierarchical level-of-detail for efficient streaming and display, we never display all of the triangles in the tileset at once; the triangle count varies from one frame to another based on the volume of space you are viewing and the size of your viewport, among other things.

Currently no public APIs exist for querying the number of triangles on screen at any given moment. Can you be more specific about what you're trying to achieve?

**Reply by peiyanlu** [reply]
Firstly, thank you for your clarification. Is it possible to calculate the cache the number of triangular faces of the Tile in the `.Tiles` file, rather than calculating the number of triangular faces of the Tile in the current viewport.

**Reply by pmconne** [reply]
It is of course possible, but no public APIs exist to do it for you. It's unclear to me why you want to do that, though. A .Tiles file is a sqlite database used to cache tiles generated by a locally-running backend. Individual tiles are generated on demand, cached, and discarded from the cache on a least-recently-used basis. The .Tiles file can contain tiles from multiple different tilesets, and never contains an exhaustive tileset representing a single iModel. So, you may draw misleading conclusions by extrapolating from metrics you obtain by counting triangles in a .Tiles file, depending on what your goal is. What is your goal?

**Reply by peiyanlu** [reply]
> It is of course possible, but no public APIs exist to do it for you. It's unclear to me why you want to do that, though. A .Tiles file is a sqlite database used to cache tiles generated by a locally-running backend. Individual tiles are generated on demand, cached, and discarded from the cache on a least-recently-used basis. The .Tiles file can contain tiles from multiple different tilesets, and never contains an exhaustive tileset representing a single iModel. So, you may draw misleading conclusions by extrapolating from metrics you obtain by counting triangles in a .Tiles file, depending on what your goal is. What is your goal?

Our potential clients need to understand the performance of iTwin, which is affected by other engines such as ThreeJs. They would like to know how many triangles the model will have after release, as well as information such as browsing frame rate and loading time, in order to evaluate the overall engine performance. We also hope to have a reliable metric to measure performance ourselves, and triangles are one of the metrics we consider, so we are trying to count the number of triangles (from. bim files or. bim. Files files), which may not be applicable to iTwin. Overall, we hope to have a reliable indicator for evaluating performance and would appreciate your professional advice.

**Reply by pmconne** [reply]
The best way to measure performance is directly - e.g., by monitoring framerate and/or memory consumption while navigating the view. That will give you real-world numbers that are directly comparable to the same metrics captured in other viewing applications. Other measurements necessarily introduce layers of indirection and can only be treated as less reliable heuristics. For iModels using HLOD 3D tiles, triangle count is an exceptionally poor heuristic for real-world performance.

**Reply by peiyanlu** [reply]
> 测量性能的最佳方法是直接测量 - 例如，通过在浏览视图时监控帧速率和/或内存消耗。这将为您提供与在其他查看应用程序中捕获的相同指标直接可比的真实数字。其他测量必然会引入间接层，只能被视为不太可靠的启发式方法。对于使用 HLOD 3D 图块的 iModel，三角形计数对于实际性能来说是一种非常糟糕的启发式方法。

Ok, thanks for the suggestion
